### ISSUE TITLE
Allow custom config to be passed into the modal template ref for use in components

### ISSUE THREAD

**Issue by wesleycho** [ng-bootstrap/ng-bootstrap#994]
Currently it is possible to be reusing the same template over again across multiple components for modals. It would be good to support passing in a config object into the modal context used for binding reusable properties.

For example, I would like to be able to do this
```html
<template #myModalTemplate let-c="close" let-d="dismiss" let-config="modalContext">
  <my-component [customProp1]="config.prop1" [customProp2]="config.prop2" (close)="c($event)" (dismiss)="d($event)"
</template>
```
in order to extract my common template & config into a reusable component.

**Comment by pkozlowski-opensource** [comment]
@wesleycho I don't think that what you are proposing is the way to go but maybe it is due to my lack of understanding of your use-case... 

Let's consider for a second that you would implement what you are proposing here. Then the code from where you use modal would look sth like:

```typescript
class MyComponent {
  modalContext = {foo: 'bar'};

  constructor(private modalService: NgbModal) {}

  open(content) {
     this.modalService.open(content, {config: modalContext});
  }
}
```

But this all is completely unnecessary since a `<template>` element is already capturing context of the host component! So we can actually do it all today without any changes to the library and with simpler code:

```typescript
class MyComponent {
  modalContext = {foo: 'bar'};

  constructor(private modalService: NgbModal) {}

  open(content) {
     this.modalService.open(content);
  }
}
```

and 

```html
<template #myModalTemplate let-c="close" let-d="dismiss">
  <my-component [customProp1]="modalContext.prop1" [customProp2]="modalContext.prop2" (close)="c($event)" (dismiss)="d($event)"
</template>
```

Alternatively if you just want to open a component as content you could simply do: 

```typescript
const modalRef = this.modalService.open(MyComponetWithContentForModal);
modalRefComponentInstance.customProp1 = modalContext.prop1;
modalRefComponentInstance.customProp2 = modalContext.prop2;
```

**Comment by pkozlowski-opensource** [comment]
@wesleycho I believe that your use-case is covered today but once again, I would like to better understand what you are trying to achieve or what you don't like about the current API. Based on my current understanding I believe that the APIs in place are actually simpler as compared to what you are proposing so it is crucial that we get good and common understanding here.

**Comment by PEsteves8** [comment]
Personally, I would also like to have the options of passing in values when opening a modal from its template reference.
For organization purposes sometimes I do prefer to have the <template> in the same place as the component that opens it, but at the same time I would prefer not to pollute the components properties with some values that are really only relevant for the modal when it opens. It's not a big deal, though.

Another issue that I had and maybe this is due to the lack of proper design or knowledge of the library,
I was going to open a modal using the ElementRef, but ended up making use of the component approach, by creating a component and passing it instead. The problem was that I wanted that component to use the same service instance as the component that triggers the opening of the modal, but I got the error 'no provider for the service', so I had to add the service to the providers array of the modal component, which felt unnecessary, since it was already provided in the module to which the main component belonged to.

**Comment by pkozlowski-opensource** [comment]
> Personally, I would also like to have the options of passing in values when opening a modal from its template reference.
For organization purposes sometimes I do prefer to have the in the same place as the component that opens it, but at the same time I would prefer not to pollute the components properties with some values that are really only relevant for the modal when it opens. It's not a big deal, though. 

We could do this but I don't think it would give you API you would love to use :-) To illustrate why let's assume that you want to open modal like so: `modalService.open(myTplRef, {data: {'foo': 'bar'}});`. Great, you want to have `foo` in your template, right. Here you go:

```html
<template let-fooInTpl="foo">
   ... {{fooInTpl}}
</template>
```

You see? You would have to _both_ pass `foo` to the `open()` method **_and_** do `let-fooInTpl="foo"` on the `<template>` element. Not great IMO.

**Comment by pkozlowski-opensource** [comment]
> Another issue that I had and maybe this is due to the lack of proper design or knowledge of the library,
I was going to open a modal using the ElementRef, but ended up making use of the component approach, by creating a component and passing it instead. The problem was that I wanted that component to use the same service instance as the component that triggers the opening of the modal, but I got the error 'no provider for the service', so I had to add the service to the providers array of the modal component, which felt unnecessary, since it was already provided in the module to which the main component belonged to.

@PEsteves8 by default we are using injector of a module as we don't have access to the component's injector in a service, I'm afraid. What you could do though is to put `ModalService` on your component level - then it will see the same injector as your component.

**Comment by pkozlowski-opensource** [comment]
@wesleycho do you feel like you still need this given https://github.com/ng-bootstrap/ng-bootstrap/issues/994#issuecomment-258893304 ? We can do this, technically speaking but this is not great API imo.

**Comment by wesleycho** [comment]
I am in complete agreement with @PEsteves8 here - sorry I took long to respond, ran a marathon and a 5K the past two days & been generally busy.

IMO it is still better than polluting a component with extraneous properties that aren't used by the component itself.

My issue is building a reusable component with bindings that are consistent so that I can DRY up my templates & thus the business logic surface area that needs testing, plus guarding against duplicate logic that may get built differently each time.

**Comment by pkozlowski-opensource** [comment]
> My issue is building a reusable component with bindings that are consistent so that I can DRY up my templates & thus the business logic surface area that needs testing, plus guarding against duplicate logic that may get built differently each time.

Why not using component as content instead of a template then?

**Comment by wesleycho** [comment]
I still have the issue of needing to bind to templates, such as displaying error messages in a consistent format.

**Comment by pkozlowski-opensource** [comment]
> I still have the issue of needing to bind to templates, such as displaying error messages in a consistent format.

I don't follow.... Sorry for being pain in a neck here but I _really_ trying to understand what would be the best API here...

**Comment by wesleycho** [comment]
For example, consider an error dialog - I want to bind an error header, and message each time I open the modal. I don't want to bind to the component which opens the modal, it makes little sense. In addition, I cannot pass the messages I get back from a failed API request in this situation. It is in this way that the current API is deficient.

**Comment by PEsteves8** [comment]
I still would like to have the option of including both the data object and the template variable.

(Note however(!), that I started learning programming intensively using Udacity one year ago, and I've been working on my first job for less than 6 months. The point is that my opinion in some situations may be purely the result of the fact that my inexperience could prevent me from realizing better ways of doing things. Just felt that that information could be important :D )

**Comment by dneuer** [comment]
> const modalRef = this.modalService.open(MyComponetWithContentForModal);
> modalRefComponentInstance.customProp1 = modalContext.prop1;

@pkozlowski-opensource isn't that racy, though? What if the component can't do anything useful without customProp1? I understand that ngOnInit is called after input properties are bound, but my assumption was that that only holds if angular creates the component through encountering it's selector in markup. Am I misunderstanding? Is angular waiting around before calling ngOnInit even for components created from within event handlers, etc?

**Comment by alex321** [comment]
@dneuer `ngOnInit` doesn't get called before this callback has finished. It is called on the next change detection cycle. This has the drawback that you have to wait if there is some initialisation that is inside `ngOnInit`.

**Comment by dneuer** [comment]
@alex321 The callback _could_ finish before the assignment to the property, no? Therefore, `ngOnInit` could be called before the assignment, too. It looks like a classic race condition, which is why people are asking for a way to pass the data into the callback itself. Or am I misunderstanding things?

**Comment by alex321** [comment]
@dneuer There are no **asynchronous** calls there so there doesn't exist any race condition.

**Comment by dneuer** [comment]
@alex321 What causes the next change detection cycle?

**Comment by dneuer** [comment]
@alex321 @pkozlowski-opensource Having read up on zones, I'm now even more convinced that this is a race:

`const modalRef = this.modalService.open(MyComponetWithContentForModal);`
`// user moves the mouse`
`modalRefComponentInstance.importantDatum1 = this.somestate;`
`// some other asynchronous  action from somewhere in my app completes`
`modalRefComponentInstance.equallyImportantDatum2 = this.someOtherState;`

If angular is kicking off change detection after every event, asynchronous action, or timeout, how is it sufficient that ModalService.open() doesn't do anything asynchronous? Apologies if I'm missing something obvious here.

**Comment by dneuer** [comment]
@alex321 @pkozlowski-opensource Ah, OK. All of this code is running in a single thread, so these calls are all sequetially executed on the stack, and then the queues are processed afterward, right? So change detection cannot run until after all the synchronous code has executed?

**Comment by alexandis** [comment]
Am I doing something wrong? Why this `roleType` me `undefined` in the template?
```
this.roleModalRef = this.modalService.open(this.roleContent, { centered: true, size: 'md' });
this.roleModalRef.componentInstance.roleType = roleType;
```
`roleType` is undefined in the template body below:
`<ng-template #roleContent let-modal let-c="close" let-d="dismiss" let-roleType="roleType">`
